### PR TITLE
Variable timestepping in Kurganov-Petrova

### PR DESCRIPTION
**Pull request by babrodtk**
Not sure if this should be merged or simply kept as a branch for posteriority. 

If we take a look at the realistic setting, we'll have H~100, dx=dy~400, and low particle velocities (u). Given the CFL condition 

dt = max_u 0.25*dx/(u+sqrt(gh))

and plugging in the numbers

dt = max_u 0.25*400/(u+sqrt(9.81)) ~ max_u 100/(u+30) 

I would assume that safely assuming u <= 5 would give a static 

dt = 100/35 = 2.85

compared to 

dt = 100/30 = 3.33

in the best case. In other words, I guess that running the extra kernel and data movement required by this PR is not offset by the increase of dt. So my gut feeling is that this PR is much more sensible to add for shock-dominated cases (which we will not focus on)

Better options are also available: it would be possible to run a somewhat pessimistic dt, and only update the timestep size every n'th iteration.

### CODE REVIEW

File: sim/src/kernels/KP07_kernel.opencl [hunk 1]
```diff
@@ -122,24 +120,46 @@ void adjustSlopes(__local float Qx[3][block_height+2][block_width+2],
...
-    
+        if (get_local_id(0) < block_height) {
+            adjustSlopeUx(Qx, RHx, Q, 1, p);
+            adjustSlopeUx(Qx, RHx, Q, block_width+2, p);
+        }
+    }    
 }
 
 
+float computeDt(const float3 Qp, const float3 Qm, float H, const float g_, const float dx) {
```
Comment (babrodtk) [inline sim/src/kernels/KP07_kernel.opencl]: This computes the timestep for a single interface

File: sim/src/kernels/KP07_kernel.opencl [hunk 2]
```diff
@@ -207,73 +237,176 @@ void init_H_with_garbage(__local float Hi[block_height+4][block_width+4],
...
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    if (nthreads >= 64) {
+        if (tid < 64 && (tid+64) < nthreads) {
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 64]);
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    
+#ifdef NVIDIA
```
Comment (babrodtk) [inline sim/src/kernels/KP07_kernel.opencl]: Note: This macro is not really well defined... Right now it's hardcoded in the common.opencl file. So for an Intel CPU e.g., you would need to remove the definition from there to get the right code path

File: sim/src/kernels/KP07_kernel.opencl [hunk 3]
```diff
@@ -207,73 +237,176 @@ void init_H_with_garbage(__local float Hi[block_height+4][block_width+4],
...
+        }
+        if (nthreads >=  4) { 
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 2]);
+        }
+        if (nthreads >=  2) { 
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 1]);
+        }
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
+#else
```
Comment (babrodtk) [inline sim/src/kernels/KP07_kernel.opencl]: The above code relies on 32-wide simd, which does not appear to work on the CPU (logically enough, just wasn't expecting it)

File: sim/src/kernels/common.opencl [hunk 4]
```diff
@@ -28,6 +28,11 @@ You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+/**
+  * Define NVIDIA pragma for assuming a 32-wide SIMD
+  */
+#define NVIDIA 1
```
Comment (babrodtk) [inline sim/src/kernels/common.opencl]: This should really be programatically set ((or detected) based on the platform being used

**Comment by setmar** [comment]
Pull request "moved" to https://github.com/babrodtk/gpu-ocean/pull/1

Will eventually be merged into variable_dt branch of metno/gpu-ocean.

**Pull request closed without merging**